clang/utils/TableGen/ClangProgModelBuiltinEmitter.cpp

```diff
@@ -848,6 +922,228 @@ void BuiltinNameEmitter::EmitQualTypeFinder() {
   OS << "\n} // Bultin2Qual\n";
 }
 
+JSONBuiltinInterfaceEmitter::TypeDesc::TypeDesc(const Record *T) : Rec(T) {
+  auto ToOCLAddressSpace = [](llvm::StringRef s) {
+    if (s == "clang::LangAS::Default")
+      return "";
+    if (s == "clang::LangAS::opencl_private")
+      return "__private";
+    if (s == "clang::LangAS::opencl_global")
+      return "__global";
+    if (s == "clang::LangAS::opencl_constant")
+      return "__constant";
+    if (s == "clang::LangAS::opencl_local")
+      return "__local";
+    return "__generic";
+  };
+
+  Name = T->getValueAsString("Name");
+  IsInteger = T->getValueAsBit("IsInteger");
+  IsSigned = T->getValueAsBit("IsSigned");
+  IsFloat = T->getValueAsBit("IsFloat");
+  Size = T->getValueAsInt("ElementSize");
+  VecWidth = T->getValueAsInt("VecWidth");
+  IsConst = T->getValueAsBit("IsConst");
+  IsVolatile = T->getValueAsBit("IsVolatile");
+  IsPointer = T->getValueAsBit("IsPointer");
+  IsBuiltin =
+      !T->isSubClassOf("FundamentalType") && !T->isSubClassOf("VectorType");
+
+  AddrSpace = ToOCLAddressSpace(T->getValueAsString("AddrSpace"));
+}
+
+std::string JSONBuiltinInterfaceEmitter::TypeDesc::GetBaseTypeAsStr() const {
+  if (Name == "void") {
+    return Name.str();
+  }
+
+  std::string TypeStr = TypePrefix;
```

Comment:
this should probably be using an ostringstream or something instead of std::string directly.

Reply:
It is now using a stream